### PR TITLE
Rewrite defineAndResources to avoid data hazard

### PR DESCRIPTION
**Pull request by KKyang**
Remove hardcoded sgpr assignments by splitting the sgpr defines into two parts. The unused defines are moved to after the kern arg load section.

### CODE REVIEW

**Review by Jinp800125 — APPROVED** [review]
LGTM

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 1]
```diff
@@ -1394,169 +1395,174 @@ def calculateWG():
...
+      with self.allocTmpSgpr(8, 2) as tmpSgpr:
+        tmpSgprM = self.sgprs["SizesFree"]
+        tmpSgprN = tmpSgprM+1
+        tmpSgprB = tmpSgprN+1
+        tmpSgprArgAddress0 = tmpSgpr.idx
+        tmpSgpr0 = tmpSgpr.idx + 2
+        tmpSgprNumWG0 = tmpSgpr.idx + 4
+        tmpSgprNumWG1 = tmpSgprNumWG0 + 1
+        tmpSgprAddrM = tmpSgprNumWG1 + 2
+        tmpSgprAccumTiles = tmpSgprAddrM + 3
```
Comment (hcman2) [inline tensilelite/Tensile/KernelWriterAssembly.py]: tmpSgprNumWG0 = tmpSgpr[4]
 tmpSgprNumWG1 = tmpSgpr[5]
 tmpSgprAddrM = tmpSgprNumWG1 + 2 = tmpSgpr[7]  <----is it a bug here?
tmpSgprAccumTiles = tmpSgprAddrM + 3 = tmpSgpr[10] <----is it a bug here?
is it possible to use tmpSgpr.idx + x for all variables?

Comment (KKyang) [reply tensilelite/Tensile/KernelWriterAssembly.py]: fixed

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 2]
```diff
@@ -1394,169 +1395,174 @@ def calculateWG():
...
+          WGMXCC = self.states.kernel["WorkGroupMappingXCC"]
+          label_skipWGMXCC = Label(label="skip_WGMXCC", comment="skip WGMXCC if no enough WGs to remap")
+          module.addComment0("only remap WGs in the range")
+          module.add(scalarStaticDivideAndRemainder(qReg=sgpr(tmpSgpr0), rReg=None, dReg=sgpr(tmpSgprNumWorkGroups), divisor=pow(WGMXCC,2), tmpSgprRes=tmpSgprRes1, doRemainder=0))
+          module.add(SMulI32(dst=sgpr(tmpSgpr0), src0=sgpr(tmpSgpr0), src1=pow(WGMXCC,2)))
+          module.add(SCmpGeU32(src0=sgpr("WorkGroup0"), src1=sgpr(tmpSgpr0)))
+          module.add(SCBranchSCC1(label_skipWGMXCC.getLabelName()))
+          module.addComment0("temp0 = (wg*%u)%%%u"%(WGMXCC,pow(WGMXCC,2)))
+          module.add(SMulI32(dst=sgpr(tmpSgpr0), src0=sgpr("WorkGroup0"), src1=WGMXCC))
+          module.add(scalarStaticRemainder(qReg=tmpSgpr, rReg=tmpSgpr0, dReg=tmpSgpr0, divisor=pow(WGMXCC,2), tmpSgprRes=tmpSgprRes1))
```
Comment (hcman2) [inline tensilelite/Tensile/KernelWriterAssembly.py:1415]: Not sure whether we can set qReg and  tmpSgprRes the same register?

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 3]
```diff
@@ -1394,169 +1395,174 @@ def calculateWG():
...
+          module.addComment0("temp1 = (wg%%%u)//%u"%(pow(WGMXCC,2), WGMXCC))
+          module.add(scalarStaticRemainder(qReg=tmpSgpr, rReg=tmpSgpr1, dReg="WorkGroup0", divisor=pow(WGMXCC,2), tmpSgprRes=tmpSgprRes1))
```
Comment (hcman2) [inline tensilelite/Tensile/KernelWriterAssembly.py:1417]: Not sure whether we can set qReg and tmpSgprRes the same register?

File: tensilelite/Tensile/KernelWriterAssembly.py (hunk 2 above)
Comment (KKyang) [reply tensilelite/Tensile/KernelWriterAssembly.py:1415]: yes cause we don't need this output

File: tensilelite/Tensile/KernelWriterAssembly.py (hunk 3 above)
Comment (KKyang) [reply tensilelite/Tensile/KernelWriterAssembly.py:1417]: yes cause we don't need this output

File: tensilelite/Tensile/KernelWriterAssembly.py [hunk 4]
```diff
@@ -1394,169 +1395,174 @@ def calculateWG():
...
+        # loop gemm count
+        module.add(SCmpLtU32(src0=sgpr(tmpSgprLoopCounter), src1=sgpr(sgprNumsOfGemm)))
+        module.add(SCBranchSCC1(labelName=label_Loop_gemm_count.getLabelName()))
+
+        # noLoadLoop
+        module.addComment1("Grouped Gemm:: noLoadLoop")
+        module.add(label_noLoadLoop)
+        module.add(SWaitCnt(lgkmcnt=0))
+        # calculate numTiles
+        regStateRes = RegisterPoolResource(idx=tmpSgpr0, size=2)
```
Comment (hcman2) [inline tensilelite/Tensile/KernelWriterAssembly.py:1503]: duplicated code with line 1479

Comment (KKyang) [reply tensilelite/Tensile/KernelWriterAssembly.py:1503]: It's originally duplicated.

**Review by hcman2 — APPROVED** [review]

**Pull request merged**